### DISCUSSION TITLE
On pictures

### DISCUSSION THREAD

**Discussion by dvv** [kanidm/kanidm#2468] [Q&A] [answered]
Hi all!

I'm considering to switch to kanidm and so far it goes well.
But I could not find any reference on whether it is possible to store person profile picture (searched issues with profile, picture, image, jpeg, avatar)?
Are there plans to cope with well known claims: https://openid.net/specs/openid-connect-basic-1_0.html#StandardClaims?

TIA

**Comment by Firstyear** [comment, marked as answer]
I think it's related to https://github.com/kanidm/kanidm/issues/2057 where we want to add in-system replicated filestorage. That would be the first step to this. Otherwise all the machinery is in place, and the claims are already present in the structures. 

We likely should turn this into an issue for avatars/display images as a follow up to #2057

**Reply by yaleman** [reply]
Yep, not yet, but it's a feature that needs developing

**Reply by yaleman** [reply]
Raised as https://github.com/kanidm/kanidm/issues/2469

**Comment by dvv** [comment]
Great news, thanks!

Another question is how do I store person' custom attributes. Say, employee's office name, address, phone, supervisor name, etc.
I found nothing on custom attributes except https://github.com/kanidm/kanidm/issues/600#issuecomment-1153279375 which is not a valid command anymore.

I think expressing hierarchy-like relations like office attendance and job supervisor _may be_ doable via custom group membership.
However the "leaf" properties of accounts and groups are left unhandled.

Please point me to the right place.

UPDATE:
I thought may be I could just use the API but no:
```
PATCH /v1/person/foo {"attrs": {"office": ["bar"]}} -> 500 {"invalidattributename":"office"}
PUT /v1/group/foo/_attr/bar ["baz"] -> 500 {"invalidattributename":"bar"}
```

**Reply by Firstyear** [reply]
You want https://github.com/kanidm/kanidm/issues/256 I think.

**Reply by Firstyear** [reply]
It won't be in the release that's about to happen, but we want this "sooner than later" as an fyi.

**Comment by dvv** [comment]
I see #256 is a long-awaited )
Am I right the it's quite a low-hanging fruit due to KV nature of our storage -- no schema needed; just open the possibility to operate on non-indexed keys?

**Reply by Firstyear** [reply]
The main thing has actually been that I wanted to do it with i18n and scim in mind. So it's low hanging but needs some care still.

**Reply by Firstyear** [reply]
Also key indexing isn't really the issue here :)

**Reply by dvv** [reply]
I mostly meant we don't need to keep them unique or enforce somehow else. They just could be a bucket of string-valued keys, and it would allow to move on, no?

**Reply by Firstyear** [reply]
Not quite so simple. kanidm entries aren't K-V pairs, but K-structures, where the structures can have rules and values. They might only allow a string, or a u32, but others allow more complex structures and multiple fields. This also has implications for how searching is done - if you search on "address" which part of the structure do we look at? The postcode? The country? So that then brings into question things like scim subattribute searching and how we handle that. 

So on the surface it looks easy, but it gets a bit "wild" once you scratch deeper.

**Reply by dvv** [reply]
Oh, I see. But can we start on pure stringy attributes for simple use cases? No even searching needed -- an authority puts the attributes, resource server gets them intact via a scope.

**Reply by Firstyear** [reply]
Post in that issue what attributes and structure you want them to have and we can decide from there :)
